addons/l10n_eg_edi_eta/tests/test_edi_json.py

```diff
@@ -0,0 +1,942 @@
+                            },
+                        ],
+                        'taxTotals': [{'taxType': 'T1', 'amount': 0.0}],
+                        'totalDiscountAmount': 3340.47878,
+                        'totalSalesAmount': 25836.91878,
+                        'netAmount': 22496.44,
+                        'totalAmount': 22496.44,
+                        'extraDiscountAmount': 0.0,
+                        'totalItemsDiscountAmount': 0.0,
+                        'signatures': [{'1': '1'}],
+                    },
+                    'response': {
+                        'l10n_eg_uuid': 'UUIDXIL9182712KMHJQ',
+                        'l10n_eg_long_id': 'LIDMN12132LASKXXA',
+                        'l10n_eg_internal_id': 'INTLA1212MMKA12',
+                        'l10n_eg_hash_key': 'BaK12lX1kASdma12',
+                        'l10n_eg_submission_id': '12125523452353',
+                    },
+                },
+            )
+
+    def test_7_simple_test_foreign_parter_exempt_discount_foreign_currency_credit_note(self):
+        with freeze_time(self.frozen_today), patch(
+            'odoo.addons.l10n_eg_edi_eta.models.account_move.AccountMove.action_post_sign_invoice',
+            new=mocked_action_post_sign_invoice,
+        ), patch(
+            'odoo.addons.l10n_eg_edi_eta.models.account_edi_format.AccountEdiFormat._l10n_eg_edi_post_invoice_web_service',
+            new=mocked_l10n_eg_edi_post_invoice_web_service,
+        ):
+            invoice = self.create_invoice(
+                move_type='out_invoice',
+                currency_id=self.currency_aed_id.id,
+                partner_id=self.partner_b.id,
+                invoice_line_ids=[
+                    {
+                        'product_id': self.product_a.id,
+                        'price_unit': 100.0,
+                        'quantity': 1.0,
+                        'product_uom_id': self.env.ref('uom.product_uom_unit').id,
+                        'discount': 10.0,
+                        'tax_ids': [(6, 0, self.env.ref(f'l10n_eg.{self.env.company.id}_eg_exempt_sale').ids)],
+                    },
+                    {
+                        'product_id': self.product_b.id,
+                        'price_unit': 100.35,
+                        'quantity': 5.0,
+                        'discount': 13.0,
+                        'product_uom_id': self.env.ref('uom.product_uom_cm').id,
+                        'tax_ids': [(6, 0, self.env.ref(f'l10n_eg.{self.env.company.id}_eg_exempt_sale').ids)],
+                    },
+                ],
+            )
+            invoice.action_post()
+            invoice.action_post_sign_invoice()
+
+            generated_files = self._process_documents_web_services(invoice, {'eg_eta'})
+            self.assertTrue(generated_files)
+            json_file = json.loads(generated_files[0])
+            self.assertEqual(
+                json_file,
+                {
+                    'request': {
```

Comment:
That is a lot of dict each time.  It is not that bad, but you could have a template that you copy and do some changes to.